### PR TITLE
Do not run all javascript after every ajax call

### LINKED ISSUE #1736: Application scripts initialization after ajax complete event
We are using turbolinks classic version 2.5.3 … maybe we have to think to upgrade to latest 5.xx stable version.

## What
Application javascript is not working as expected because inconsistency on ajax complete event callbacks.

It seems that remote calls managed by Turbolinks are throwing ajaxComplete event while ajax calls managed directly by jQuery are throwing ajax:complete event. This situation make developers to lose time to understand how application javascript behave and forces them to create unnecessary code inside scripts to make application work as they expect.

## Where
At main application scripts manifest we can find these following lines:

```
$(function(){
  Turbolinks.enableProgressBar()

  // This line initialize application for direct page loading (not through Turbolinks)
  $(document).ready(initialize_modules);
  // This line initialize application for pages loaded by Turbolinks
  $(document).on('page:load', initialize_modules);
  // This line tries to initialize application modules after complete ajax calls but ajax calls 
  // managed by turbolinks is throwing ajaxComplete so this line is not executed
  $(document).on('ajax:complete', initialize_modules);
});
```
Due to this behavior some AJAX responses have to call application modules manually to initialize new DOM elements created.

When we change definition to be turbolinks compliant (ajaxComplete)
```
 $(document).on('ajaxComplete', initialize_modules);
```
… some collateral damage arises, i’ll try to explain with proposal pages as example, that have many ajax components like followable, annotatable, flaggable,

1 - User clicks on flag to mark proposal as offensive, this launches an Turbolinks managed call, this call throws ajaxComplete event when finished
2- This ajaxComplete callback call all application modules initialization again
3- Now, all application modules are initialized again and the javascript behaviour is duplicated

## How
I think it would be nice to have only one of those ajax callbacks working in both situations and avoid unnecessary DOM inititalizations.

**Solution 1: Manual initialization**
1. Remove this line from application js manifest 
```$(document).on('ajax:complete', initialize_modules);```
2. Initialize only modules needed at any ajax response manually. If you are refreshing comments module we should initialize only comments module at the same response, there is no need to call all application modules initialization, we only need to initialize new DOM elements.

**Solution 2: Fix current configuration**
1. Standarize one ajax complete callback for all application.
2. Review all application javascript modules to ensure they are not initializing already initialized DOM elements.

### PR DESCRIPTION
**Pull request by Senen**
## References

* Closes #1736 

## Objectives
Run application javascript initialization once and update affected javascript modules to work with this new pattern.

Now that we are not initializing all the DOM after every ajax call we must ensure that new elements added through AJAX (only those that need javascript to work) are initialized as expected by using  "Delegated Event Handlers" [1] when possible or by doing it manually at `js.erb` file.

[1] https://api.jquery.com/on/#direct-and-delegated-events

## Visual Changes
None

## Notes
Successfully tested on production environment!

As this kind of change feels dangerous we had to review all places where using ajax calls to ensure elements added to the DOM were being initialized as we expect.

### CODE REVIEW

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -173,5 +173,4 @@ $(function() {
 
   $(document).ready(initialize_modules);
   $(document).on("page:load", initialize_modules);
-  $(document).on("ajax:complete", initialize_modules);
```
[the remark this answers is not in the document]
Comment (Senen) [reply app/assets/javascripts/application.js:176]: Yes, that's right, probably i can reproduce that wrong behavior on a new spec.

File: spec/system/comments/debates_spec.rb [hunk 2]
```diff
@@ -230,6 +230,26 @@
     expect(page).not_to have_selector("#js-comment-form-comment_#{comment.id}", visible: true)
   end
 
+  scenario "Reply to reply", :js do
+    create(:comment, commentable: debate, body: "Any estimates?")
+    login_as(create(:user))
+    visit debate_path(debate)
```
[the remark this answers is not in the document]
Comment (Senen) [reply spec/system/comments/debates_spec.rb:256]: Done!

File: app/assets/javascripts/application.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Senen) [reply app/assets/javascripts/application.js:176]: Finally i was able to add a spec to check the behavior of comments collapse feature. 👌

File: spec/system/comments/debates_spec.rb [hunk 3]
```diff
@@ -84,6 +84,25 @@
     expect(page).not_to have_content grandchild_comment.body
   end
 
+  scenario "can collapse comments after adding a reply", :js do
+    parent_comment = create(:comment, body: "Main comment", commentable: debate)
+    create(:comment, body: "First subcomment", commentable: debate, parent: parent_comment)
+
+    login_as(user)
+    visit debate_path(debate)
+    within "#comment_#{parent_comment.id}_reply" do
```
Comment (javierm) [inline spec/system/comments/debates_spec.rb:93]: We usually leave a blank line before and after `within` blocks. Sorry to insist on this one :smile:.

**Comment by Senen** [comment]
Hi @javierm, thank for your review. I think i fixed everything you ask. Let me know if i missed some "air"

**Pull request merged**